src/main/java/io/cdap/plugin/sap/transformer/ODataEntryToRecordTransformer.java

```diff
@@ -131,13 +146,208 @@ private Object extractValue(String fieldName, Object value, Schema schema) {
+          Schema multiPolygonSchema = multiPolygonsField.getSchema().getComponentSchema();
+          multiPolygons.add(extractGeospatial(fieldName, g, multiPolygonSchema));
+          break;
+      }
+    });
+
+    return StructuredRecord.builder(schema)
+      .set(SapODataConstants.GEOSPATIAL_TYPE_FIELD_NAME, "GeometryCollection")
+      .set(SapODataConstants.GEOSPATIAL_DIMENSION_FIELD_NAME, collection.getDimension().name())
+      .set(SapODataConstants.GEO_COLLECTION_POINTS_FIELD_NAME, points)
+      .set(SapODataConstants.GEO_COLLECTION_LINE_STRINGS_FIELD_NAME, lineStrings)
+      .set(SapODataConstants.GEO_COLLECTION_POLYGONS_FIELD_NAME, polygons)
+      .set(SapODataConstants.GEO_COLLECTION_MULTI_POINTS_FIELD_NAME, multiPoints)
+      .set(SapODataConstants.GEO_COLLECTION_MULTI_LINE_STRINGS_FIELD_NAME, multiLineStrings)
+      .set(SapODataConstants.GEO_COLLECTION_MULTI_POLYGONS_FIELD_NAME, multiPolygons)
+      .build();
+  }
+
+  private StructuredRecord extractPoint(Point point, Schema schema) {
+    return StructuredRecord.builder(schema)
+      .set(SapODataConstants.GEOSPATIAL_DIMENSION_FIELD_NAME, point.getDimension().name())
+      .set(SapODataConstants.POINT_X_FIELD_NAME, point.getX())
+      .set(SapODataConstants.POINT_Y_FIELD_NAME, point.getY())
+      .set(SapODataConstants.POINT_Z_FIELD_NAME, point.getZ())
+      .build();
+  }
+
+  private StructuredRecord extractLineString(LineString lineString, Schema schema) {
+    Schema.Field coordinatesField = schema.getField(SapODataConstants.GEOSPATIAL_COORDINATES_FIELD_NAME);
+    Schema pointSchema = coordinatesField.getSchema().getComponentSchema();
+    List<StructuredRecord> coordinates = new ArrayList<>();
+    Iterator<Point> pointIterator = lineString.iterator();
+    if (pointIterator != null) {
+      pointIterator.forEachRemaining(p -> coordinates.add(extractPoint(p, pointSchema)));
+    }
+    // "LineString" and "MultiPoint" schemas are the same. Type name is required to distinguish them
+    return StructuredRecord.builder(schema)
+      .set(SapODataConstants.GEOSPATIAL_TYPE_FIELD_NAME, "LineString")
+      .set(SapODataConstants.GEOSPATIAL_DIMENSION_FIELD_NAME, lineString.getDimension().name())
+      .set(SapODataConstants.GEOSPATIAL_COORDINATES_FIELD_NAME, coordinates)
+      .build();
+  }
+
+  private StructuredRecord extractPolygon(Polygon polygon, Schema schema) {
+    Schema.Field exteriorField = schema.getField(SapODataConstants.POLYGON_EXTERIOR_FIELD_NAME);
+    Schema pointSchema = exteriorField.getSchema().getComponentSchema();
+    List<StructuredRecord> exterior = new ArrayList<>();
+    if (polygon.getExterior() != null && polygon.getExterior().iterator() != null) {
+      polygon.getExterior().iterator().forEachRemaining(point -> exterior.add(extractPoint(point, pointSchema)));
+    }
+    List<StructuredRecord> interior = new ArrayList<>();
+    for (int i = 0; i < polygon.getNumberOfInteriorRings(); i++) {
```

Comment:
how come we're flattening the interior rings instead of preserving the fact that Polygon has a collection of collection of Points as the interior?

Reply:
I did it by mistake because of examples with a single interior ring.

There are actually can be multiple rings. From The [GeoJSON Format](https://tools.ietf.org/html/rfc7946#section-3.1.6):
> For Polygons with more than one of these rings, the first MUST be
  the exterior ring, and any others MUST be interior rings.  The
  exterior ring bounds the surface, and the interior rings (if
  present) bound holes within the surface. 

"interior" mapping changed to be array of `LineString` records as in the [Olingo's Java AP](https://github.com/apache/olingo-odata4/blob/master/lib/commons-api/src/main/java/org/apache/olingo/commons/api/edm/geo/Polygon.java#L31)I: 
```
...
ComposedGeospatial<LineString> interiorRings;
...

```
Thanks.